README.md

```diff
@@ -0,0 +1,139 @@
+# mobx-state-tree typescript slowness & duplications in `d.ts` emit (And how to make it better) [part 1]
+
+> This article contains work, knowledge and experience accumulated by others and myself regrading improving the performance of typescript build, edit and IDE responsiveness.  
+You can find almost all this information scattered in github issues, comments and code snippets.  
+I will try to provide references, citations and bibliography. Please help me with that :)
+
+`mobx-state-tree` is an interesting creation that takes types to a new level and typescript inference power to it's limits.  
+It's doing so by defining runtime models and types in code, and inferring design-time typescript informatio.
+On this series of articles, we will learn about some caveats and how to overcome them.
+
+
+
+## The problem(s)
+Emitting typescript types for naive `MST` code, that uses any form of compositing,
+will lead to duplication/inline of the entire tree of model types on each reference point.
+
+Example:
+```ts
+export const Store1 = t
+    .model('Store1', {
+        store2: t.optional(t.array(Store2), [])
+    });
+```
+<sub>based on https://github.com/garygentry/mst-decl-test/blob/master/src/stores/store1.ts</sub>
+
+Will emit:
+
+```ts
+export declare const Store1: import("mobx-state-tree").IModelType<{
+    store2: import("mobx-state-tree").IOptionalIType<import("mobx-state-tree").IArrayType<import("mobx-state-tree").IModelType<{
+        store3: import("mobx-state-tree").IArrayType<import("mobx-state-tree").IModelType<{
+            store4: import("mobx-state-tree").IModelType<{
+                store5: import("mobx-state-tree").IModelType<{
+                    prop1: import("mobx-state-tree").IMaybe<import("mobx-state-tree").ISimpleType<string>>;
+                }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>;
+            }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>;
+            store5: import("mobx-state-tree").IOptionalIType<import("mobx-state-tree").IModelType<{
+                prop1: import("mobx-state-tree").IMaybe<import("mobx-state-tree").ISimpleType<string>>;
+            }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>, [undefined]>;
+        }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>>;
+    }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>>, [undefined]>;
+}, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>;
+```
+What we can see here is that instead of referencing types of the child models, we get full duplication/inlining of child models declarations.
```

Comment:
```suggestion
What we can see here is that instead of referencing types of the child models, we get full duplication/inlining of child model declarations.
```